[IMP] l10n-italy: partner bank check on invoices or sale orders with RiBa payment mode

Chionne27 opened the pull request.

SirPyTech commented on l10n_it_fatturapa_out_riba/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "auto_install": True,
+    "depends": [
+        "l10n_it_fatturapa_out",
+        "sale",
```
Se dipende da `l10n_it_fatturapa_out` e `sale`, perché chiamarlo `l10n_it_fatturapa_out_riba` e non `l10n_it_fatturapa_out_sale`?
Se poi le dipendenze si potessero ridurre ad essere `l10n_it_fatturapa` e `sale`, queste modifiche si potrebbero aggiungere al modulo esistente `l10n_it_fatturapa_sale`, cosa ne pensi?

SirPyTech replied:
:+1:

SirPyTech commented on l10n_it_fatturapa_out_riba/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "name": "ITA - Controllo sulla banca ricevente per pagamento RiBa",
+    "summary": """ This module avoids to add a partner bank when a RiBa
+        payment mode is set up on customer invoices.""",
+    "version": "14.0.0.0.0",
```
```suggestion
    "version": "14.0.1.0.0",
```
Vedi https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#12version-numbers:
> The version number in the module manifest should be the Odoo major version (e.g. 12.0) followed by the module x.y.z version numbers. For example: 12.0.1.0.0 is expected for the first release of an 12.0 module.

SirPyTech replied:
:+1:

SirPyTech commented on l10n_it_fatturapa_out_riba/models/account.py:
```diff
@@ -0,0 +1,22 @@
+            if move.move_type == "out_invoice" and move.invoice_payment_term_id:
+                # MP12 refers to RIBA payment mode
+                if (
+                    move.invoice_payment_term_id.fatturapa_pm_id.code == "MP12"
```
Questo controllo non ha bisogno del modulo delle RiBa né di `sale` per funzionare quindi si potrebbe mettere direttamente in `l10n_it_fatturapa_out` o addirittura in `l10n_it_fatturapa`, cosa ne pensi?

SirPyTech replied:
:+1:

SirPyTech commented on l10n_it_fatturapa_out_riba/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,6 @@
+This module avoids to add a partner bank when a RiBa payment mode is set up on customer invoices.
+This allows to create electronic invoices without a partner bank name and it's IBAN. If an invoice
+is created by a sale order and the payment mode is RiBa, the partner bank will autmatically deleted
```
```suggestion
is created by a sale order and the payment mode is RiBa, the partner bank will automatically deleted
```
Potresti anche aggiungere la descrizione in italiano come per gli altri moduli del repository? Vedi ad esempio https://github.com/OCA/l10n-italy/blob/073505eae5b16108a871057465f2a8d73e69ff99/l10n_it_fatturapa_sale/readme/DESCRIPTION.rst?plain=1#L1-L7

SirPyTech replied:
Puoi aggiungere della documentazione su questa nuova feature nei moduli che vengono modificati?

Credo sia importante spiegare il motivo di questa modifica da qualche parte, nel codice o nel commit o nel README.
Perché altrimenti ci ritroviamo con del codice che invalida delle fatture perché "C'è il termine di pagamento MP12 e il conto bancario del cliente": ma perché questo tipo di fattura non va bene?
E se nessuno se lo ricorda è difficilmente mantenibile, senza dei test poi è ancora più facile che la modifica si perda col tempo.

Giustamente l'hai spiegato nella issue https://github.com/OCA/l10n-italy/issues/4814:
> Quando viene fatta la fattura non si sa mai quale sarà la banca ricevente delle riba perchè viene deciso in una fase successiva. Quindi mettendo la banca ricevente all'emissione fattura potrebbe trarre in inganno il cliente e magari creare contestazioni

ma io la vedo come una modifica così criptica che meglio motivarla anche in qualche punto più vicino al codice.

SirPyTech replied:
:+1:

SirPyTech commented on l10n_it_fatturapa_out_riba/tests/test_fatturapa_out_riba.py:
```diff
@@ -0,0 +1,108 @@
+
+
+class TestFatturapaOutRiba(TransactionCase):
+    def setUp(self, *args, **kwargs):
```
Si potrebbe usare invece `SavepointCase` e `setUpClass`? Così il setup viene fatto una sola volta e velocizziamo l'esecuzione dei test.

SirPyTech replied:
Come mai non ci sono i test nelle ultime modifiche? È possibile ripristinarli?

Chionne27 replied:
Li ho ripristinati

SirPyTech replied:
:+1:

SirPyTech reviewed (changes requested):
Grazie della PR!
Ho fatto revisione del codice e trovato i punti qui sotto.
Il problema di fondo è che in realtà queste modifiche a livello tecnico non hanno a che fare con il modulo delle RiBa (e solo in parte con `sale`), ma si vuole semplicemente impedire che un particolare metodo di pagamento (MP12) possa essere impostato sulle fatture che hanno un conto impostato.

Il controllo sulla fattura potrebbe andare in `l10n_it_fatturapa`, mentre il controllo sull'ordine di vendita potrebbe andare in `l10n_it_fatturapa_sale`, senza bisogno di creare nuovi moduli.
Idealmente il controllo di MP12 andrebbe implementato una volta sola e chiamato dove opportuno, ma è talmente poca roba che si potrebbe anche rimandare a un refactoring successivo.

---

Potresti creare una issue che spieghi qual è il problema/feature che queste modifiche risolvono/implementano? Vedi https://github.com/OCA/l10n-italy/wiki/Team-di-sviluppo#apertura-issue.

La issue aiuterebbe a capire come mai andrebbe sollevata un'eccezione quando le fatture con MP12 hanno il conto del partner, o addirittura andrebbe svuotato il conto del partner nelle fatture generate da ordini di vendita che hanno un termine di pagamento che ha un metodo di pagamento con codice MP12 :face_with_spiral_eyes:.
Se fosse un po' più chiara la motivazione dietro tutto questo, forse anche altri revisori verrebbero a dare un'occhiata.

Chionne27 committed:
[IMP] l10n_it_fatturapa: check RiBa Partner bank

Chionne27 committed:
[IMP] l10n_it_fatturapa_sale: Remove partner bank for RIBA payments

Clear partner bank recipient when a sale order is set with RiBa payment mode before invoice create

SirPyTech commented on l10n_it_fatturapa/models/account.py:
```diff
@@ -459,3 +460,19 @@ def open_form_current(self):
+                    raise ValidationError(
+                        _(
+                            "It is not possible to set a partner "
+                            "bank with RIBA in payment mode for custmer invoices."
```
> Non è possibile impostare la banca di un partner con RiBa nella modalità di pagamento per le fatture cliente.

Non capisco, e credo che nemmeno l'utente finale capirebbe; dici che si riesce a scrivere un po' meglio?

Sarebbe utile anche aggiungere delle istruzioni su come passare questa validazione, ad esempio:
- Rimuovere la banca del partner
- Cambiare il termine di pagamento

Chionne27 replied:
Come mai non si capisce? Io intendo proprio che non si può impostare una banca ricevente (il campo si chiama partner_bank_id) e il modo di pagamento RIBA insieme. 
Hai ragione sul dare istruzioni

SirPyTech replied:
Rileggendolo un po' di volte e sapendo tutto il contesto si capisce, ma mettendomi nei panni dell'utente non so quanto lo capirei, ma capisco che è soggettivo.
Le istruzioni comunque aiutano quindi per me è ok :+1:

SirPyTech commented on l10n_it_fatturapa/models/account.py:
```diff
@@ -459,3 +460,19 @@ def open_form_current(self):
+                    raise ValidationError(
+                        _(
+                            "It is not possible to set a partner "
+                            "bank with RIBA in payment mode for custmer invoices."
```
```suggestion
                            "bank with RIBA in payment mode for customer invoices."
```

SirPyTech replied:
:+1:

SirPyTech commented on l10n_it_fatturapa/models/account.py:
```diff
@@ -459,3 +460,19 @@ def open_form_current(self):
         }
+
+    @api.constrains("move_type", "invoice_payment_term_id", "partner_bank_id")
+    def _check_riba_partner_bank(self):
```
C'erano dei test per questa funzionalità, come mai sono spariti?

SirPyTech replied:
👍

SirPyTech commented on l10n_it_fatturapa_sale/models/sale_order.py:
Il titolo di questo commit risulta tagliato: 
<img width="2500" height="161" alt="image" src="https://github.com/user-attachments/assets/9748380a-b583-4051-a056-7a4bad81214e" />
lo potresti accorciare?
Vedi https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message:
> When you open a PR, please check if the commit message is cut with ellipsis.
> When this happens, it means your message is too long. You should shorten it. Start by rephrasing and keeping the summary very synthetic. The explanation or motivation should be kept in the description of the commit.

SirPyTech replied:
👍

SirPyTech reviewed (changes requested):
Grazie di aver creato la issue, vedi i punti ancora aperti qui sotto.

SirPyTech commented on l10n_it_fatturapa_sale/models/sale_order.py:
```diff
@@ -38,3 +38,20 @@ def unlink(self):
+
+    def _prepare_invoice(self):
+        """
+        English:
```
```suggestion
```

SirPyTech commented on l10n_it_fatturapa/__init__.py:
```diff
@@ -3,3 +3,4 @@
 
 from . import models
 from . import controllers
+from . import tests
```
**note**: Questo non serve, si potrebbe anche rimuovere

SirPyTech reviewed (approved):
Un paio di note ma niente di bloccante

Marco-ooops reviewed (approved):
test funzionale ok

francesco-ooops commented:
/ocabot merge minor
